[macOS] Support rendering native <progress> in vertical writing mode

pxlcoder opened the pull request:
#### e105446b239e5bea36bb218c7392271e850da146
<pre>
[macOS] Support rendering native &lt;progress&gt; in vertical writing mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=261493">https://bugs.webkit.org/show_bug.cgi?id=261493</a>
rdar://115408537

Reviewed by Tim Nguyen and Richard Robinson.

Add support for rendering `&lt;progress&gt;` with `appearance: auto` with a vertical
writing mode, by rotating the CoreUI image by 90 degrees. This approach is viable
since the image does not have any shadows or orientation specific behaviors.

* LayoutTests/fast/forms/vertical-writing-mode/progress-expected.html: Added.
* LayoutTests/fast/forms/vertical-writing-mode/progress.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/progress-appearance-native-computed-style.optional-expected.txt:

These subtests were previously incorrectly passing, as the minimum block size
was not actually being enforced with a vertical writing mode. Now, the vertical
writing mode fails the same way as the horizontal writing mode.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
(WebCore::ProgressBarMac::rectForBounds const):

Adjust `rectForBounds` to ensure a minimum block-size is enforced. Support for
vertical writing modes is not added at a &quot;lower-level&quot; (`controlSizeForFont`,
`cellSize`, `cellOutsets`), since all controls can not necessarily be rotated.

(WebCore::ProgressBarMac::draw):

Draw vertical progress bars into a horizontal buffer and rotate the context to
achieve a vertical appearance.

Canonical link: <a href="https://commits.webkit.org/267988@main">https://commits.webkit.org/267988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b68710af511e74fc73e0a4f32ef0629cd1ddee12

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18263 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19174 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20102 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17092 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18458 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18750 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19028 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18484 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18694 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20985 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15917 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16655 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23150 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16936 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16826 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21039 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17390 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14749 "7 flakes 1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16488 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16494 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4359 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20850 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17235 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (changes requested):
The fact that we don't see a  progression on these tests: https://wpt.fyi/results/css/css-writing-modes/forms?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2022-forms%20or%20label%3Ainterop-2023-forms%20progress-appearance

likely means the test needs work. Can we add WPT coverage?

Also can you attach screenshots of how that looks with direction: rtl? A screenshot of https://bug-245355-attachments.webkit.org/attachment.cgi?id=463750 would be ideal.

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -154,6 +154,9 @@ editing/pasteboard/paste-dark-mode-color-filtered.html [ Skip ]
 fast/loader/plain-text-document-dark-mode.html [ Skip ]
 
+# Only applicable on platforms with vertical form control support
+fast/forms/vertical-writing-mode/progress.html [ Skip ]
```
I'd rather blocklist platforms that don't have support rather than allowlist platforms that do.

pxlcoder replied:
Done.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/progress-appearance-native-computed-style.optional-expected.txt:
```diff
@@ -1,6 +1,6 @@
 
 FAIL progress[style="writing-mode: horizontal-tb"] block size should match height and inline should match width assert_equals: expected "16px" but got "21px"
-PASS progress[style="writing-mode: vertical-lr"] block size should match width and inline should match width
-PASS progress[style="writing-mode: vertical-rl"] block size should match width and inline should match width
+FAIL progress[style="writing-mode: vertical-lr"] block size should match width and inline should match width assert_equals: expected "16px" but got "21px"
+FAIL progress[style="writing-mode: vertical-rl"] block size should match width and inline should match width assert_equals: expected "16px" but got "21px"
```
Is there a reason we inconsistently apply the minimum block size?

pxlcoder replied:
The test specifically has `assert_equals(style.blockSize, "16px");`, which seems unnecessary. We pass all of the important assertions:

```
assert_equals(style.blockSize, style.height);
assert_equals(style.inlineSize, style.width);
```

and 

```
assert_equals(style.blockSize, style.width);
assert_equals(style.inlineSize, style.height);
```

It was previously "passing" for `vertical-*`, since our native rendering code was not enforcing our default block size of 21px. We should probably remove the explicit size assertion from the test.

nt1m replied:
I think the intent of that 16px assertion is to test the aspect ratio. Can we adjust the test so it tests the aspect ratio in a smarter way? (E.g. inline-size wider than block-size)

pxlcoder replied:
Sounds good – I'll do that tomorrow in a separate PR.

nt1m commented on Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
```diff
@@ -67,16 +67,28 @@
 {
-    int minimumProgressBarHeight = sizeForSystemFont(style).height();
-    if (bounds.height() > minimumProgressBarHeight)
+    auto isVerticalWritingMode = style.states.contains(ControlStyle::State::VerticalWritingMode);
```
Could just be a `bool`

pxlcoder replied:
Leaving this as `auto`, since that's what we generally use in new code, and it's clear from both the left and right hand sides that this is a `bool`.

nt1m commented on Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
```diff
@@ -126,6 +143,13 @@
 
     GraphicsContextStateSaver stateSaver(context);
 
+    if (isVerticalWritingMode) {
+        context.translate(inflatedRect.height(), 0);
+        context.translate(inflatedRect.location());
+        context.rotate(piOverTwoFloat);
+        context.translate(-inflatedRect.location());
+    }
```
This is fine for now, but I assume we'll want some kind of shared helper so we don't duplicate this between different controls.

pxlcoder replied:
Let's see, right now I think `<meter>` going to be the only other control which will benefit from this exact logic.

pxlcoder committed:
[macOS] Support rendering native <progress> in vertical writing mode
https://bugs.webkit.org/show_bug.cgi?id=261493
rdar://115408537

Reviewed by Tim Nguyen and Richard Robinson.

Add support for rendering `<progress>` with `appearance: auto` with a vertical
writing mode, by rotating the CoreUI image by 90 degrees. This approach is viable
since the image does not have any shadows or orientation specific behaviors.

* LayoutTests/fast/forms/vertical-writing-mode/progress-expected.html: Added.
* LayoutTests/fast/forms/vertical-writing-mode/progress.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-writing-modes/forms/progress-appearance-native-computed-style.optional-expected.txt:

These subtests were previously incorrectly passing, as the minimum block size
was not actually being enforced with a vertical writing mode. Now, the vertical
writing mode fails the same way as the horizontal writing mode.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
(WebCore::ProgressBarMac::rectForBounds const):

Adjust `rectForBounds` to ensure a minimum block-size is enforced. Support for
vertical writing modes is not added at a "lower-level" (`controlSizeForFont`,
`cellSize`, `cellOutsets`), since all controls can not necessarily be rotated.

(WebCore::ProgressBarMac::draw):

Draw vertical progress bars into a horizontal buffer and rotate the context to
achieve a vertical appearance.

Canonical link: https://commits.webkit.org/267988@main
